### ISSUE TITLE
Sonic Rivals 2 too bright in Buffered Mode (now too slow)

### ISSUE THREAD

**Issue by FireChaos** [hrydgard/ppsspp#6117]
![untitled](https://cloud.githubusercontent.com/assets/7671689/3058877/1eebf3e6-e1e5-11e3-92b2-b999c84fc600.png)
This happens when I'm playing this game. it is not normal and it bothers me because I can't see where I'm stepping at. Any help?

**Comment by unknownbrackets** [comment]
I don't really know what this game is doing, but it's doing something very, very, very evil.

Step 1. Break the screen into a grid of blocks that are sized 32x34 (so, 15x8 blocks, or 120 total blocks.)
Step 2. Somehow render a scene (I think) to a temporary buffer, at regular 480x272 size.
Step 3. Go through each block in the grid, from top left going down and then right:
Step 3.1. First, depal the temp buffer to get red (CLUT32 shift 0).  Draw this to the screen with no blending or tests - so just draw it.
Step 3.2. Next, same block, extract the green (shift is 8 now. ) This time, turn on blending (ONE + ONE), but leave everything else off still.
Step 3.3. Don't forget about blue.  Can't move on to the next block without shift=16.  Still blending as before.
Step 3.4. Copy some portion (using texturing) of our same source temp buffer (this time as 8888) to another temporary buffer.  I'm not sure if this is the same block but it's done every other (?) block anyway (the above are all done every single block.)
Step 4. Apply bloom from the result (not sure exactly how it does this.)

So, in other words, rendering a frame in this game currently involves exactly 120 * 3 = 360 separate depal passes.  All except the first 3 just regenerate data we already created in previous passes.  It also swaps framebuffers at least 60 times during this, and uses the framebuffer as a non-depal texture that many times too.

To recap: this game does something very, very, very evil.  I mean, why would you do that?  Just to be evil, right?  It can't be necessary.  No part of that horror is necessary.

-[Unknown]

**Comment by hrydgard** [comment]
They clearly want to use depal to apply some gamma curve or similar separately on R, G and B, for what purpose I do not know, to make the bloom look better or something. It's also known that there is some kind of performance issue with PSP VRAM that usually makes it beneficial to do things that fill a lot of pixels in 32 or 64-pixel wide stripes (this can be seen in the strange multi-rect clears that games do). But I don't get why it would split vertically, and why it wouldn't just do these things as multi-rect draws.. and all that render target swapping...

**Comment by unknownbrackets** [comment]
Right.  I get that it's doing different adjustments for R, G, and B, but it could do a bunch of rects all at once (it's actually using triangle strips) to get the job done from the same texture with the same clut.  I'm pretty sure this would perform better than what they're doing now even on a real PSP...

I guess our best option would be caching X recent depal FBOs per framebuffer?  But that could really drink up VRAM...

-[Unknown]

**Comment by hrydgard** [comment]
Or just throw up our hands and optimize the software rasterizer :) Yeah, I dunno ...

**Comment by unknownbrackets** [comment]
Another idea is just bypassing the blit entirely for certain depals.

For example, we could dedicate 2-3 bits of the frag shader id to a "depal profile id", or simply a "clean" swizzle.  If the value != 0, we would do the depal directly in the normal rendering fragment shader.

If we ran out of profile id space, we'd continue depaling the current way, and we might do it the current way still for certain complex depals.  But since this game is only using a certain full channel (r, g, and b), it should be relatively simple to do this way.

And this would mean a lot less copying.  Though, I'm not sure what the speed impact would be to games that use depal more naturally...

-[Unknown]

**Comment by ppmeis** [comment]
Tested with latest build. Everything is too slow but with good color using 10x IR:
![image](https://cloud.githubusercontent.com/assets/4381277/8891943/67fea4e8-333d-11e5-94ae-1ed1ba17ee74.png)

Using 4x IR game runs a good speed and colors:
![image](https://cloud.githubusercontent.com/assets/4381277/8891950/d234623a-333d-11e5-9bc8-33ec9cb75f16.png)

But if I enable "Disable slower effects", speed is normal but bright is too high:
![image](https://cloud.githubusercontent.com/assets/4381277/8891937/47502a8c-333d-11e5-9df8-06135cd62811.png)

**Comment by TKC2802** [comment]
How do I fix this on Android? I tried to play sonic rivals 2, but when I start the race, the whole thing just lags badly and the screen turns very bright. This is in buffered rendering mode. When I changed to non-buffered rendering mode, the whole thing ran smoothly. However, the screen turned totally black instead. I am using redmi 1s.

**Comment by hrydgard** [comment]
@TKC2802  you don't, there's currently no way to run this game properly with good performance. It's doing things that are completely insane on modern mobile GPUs but happened to run fast enough on the PSP's.

**Comment by Mann-Nova** [comment]
So has anyone figured out how to make this game run smoothly at any resolution higher than 1x? Like previously mentioned, Disabling Slower Effects causes the game to have some weird bloomed up brightness that makes it rather unenjoyable to play, but I can't think of another way to run it (and Rivals 1, for that matter) at a high resolution smoothly. Really unfortunate for someone who needs to capture game footage in HD and can't.

**Comment by unknownbrackets** [comment]
Possibly now still too bright in Vulkan.  Might help if someone gave comparison screenshots in each backend.

Vulkan may run this game smoother now.

-[Unknown]

**Comment by mgemt** [comment]
I Confirm on the latest android build colors has been fixed, the games (Sonic Rivals 1) run perfect 30 fps with Vulkan backend on Android v1.7.5-414 but Struggle with the OpenGL ES Backend
Vulkan:
![screenshot_20190206-200651](https://user-images.githubusercontent.com/44252828/52343859-dff80d00-2a4b-11e9-943c-b8efcc4eccb3.png)
Android:
![screenshot_20190206-200725](https://user-images.githubusercontent.com/44252828/52343871-e9817500-2a4b-11e9-8a80-a601c09a6a34.png)

**Comment by hrydgard** [comment]
Yeah, Vulkan runs this game great now after I added the hack that rearranges the frame rendering operations for performance on modern GPUs.  Unfortunately haven't implemented this yet for OpenGL, but it works at least, the only remaining issue is that perf could be improved so I'm closing this.

**Comment by Duglor** [comment]
So i have this brightness problems on sonic 1 and sonic 2 on lr-ppsspp for retropie. What must i change any retroarch settings to in order to fix this issue?

**Comment by hrydgard** [comment]
Can you show a screenshot of your problem?

**Comment by Duglor** [comment]
Having problems with my pc currently...so in the meantime, it's basically this exactly. I mean it looks exactly like these to the T. 

![Image1](https://github.com/hrydgard/ppsspp/assets/53085191/2070f432-ffd5-4fae-950b-e4a9c0d581c8)
![Image2](https://github.com/hrydgard/ppsspp/assets/53085191/aa8eed85-d9dc-4cd4-9804-248a7d8db697)

**Comment by placeboyue** [comment]
This happened to me on an actual PSP. Surprised to barely see any people talking about the issue
